### PR TITLE
[v0.31] feat/bitfinex connector parts 1-4

### PR DESCRIPTION
**Pull request by nionis**
**Before submitting this PR, please make sure**:
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/", etc)

**Optional**:
- Related Github issue:
- Related Clubhouse Story:

**A description of the changes proposed in the pull request**:
Implementation of the Bitfinex connector

**TODO**
- [ ] intermittent connection issue
- [ ] cosmetic bug with the end price in history
- [ ] setting minimum order size sometimes doesn't work
- [ ] markets like (EOS-ETH, ZRX-ETH) mid price is wrong
- [ ] update documentation

### CODE REVIEW

**Comment by dennisocana** [comment]
Couldn't place orders through Hummingbot even if above the minimum order size.

putty_ZQfDq0Si6b

My `order_amount` is `0.001` and the minimum order size for BTC-USD pair is `0.0006`.

```
"pair": "btcusd",
"price_precision": 5,
"initial_margin": "20.0",
"minimum_margin": "10.0",
"maximum_order_size": "2000.0",
"minimum_order_size": "0.0006",
"expiration": "NA",
"margin": true
```

I can place limit orders in Bitfinex manually with the same amount just fine.

chrome_LWFAOMyILM

Log file : [bitfinex_order_amount.log](https://github.com/CoinAlpha/hummingbot/files/4728089/bitfinex_order_amount.log)

**Comment by dennisocana** [comment]
All unit tests are passing

```
test_cancel_all (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_execute_limit_buy (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_execute_limit_sell (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_get_balance (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_get_fee (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_limit_buy (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_limit_sell (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_minimum_order_size (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok
test_orders_saving_and_restoration (test.integration.test_bitfinex_market.BitfinexMarketUnitTest) ... ok

----------------------------------------------------------------------
Ran 9 tests in 57.104s

OK
```

**Comment by dennisocana** [comment]
Found a bug where it creates alternating single order on one side only (buy -> sell -> buy -> sell) when inventory skew is enabled.

No errors found in logs : [alternating_orders.log](https://github.com/CoinAlpha/hummingbot/files/4734467/alternating_orders.log)

FwyuTI48VW

**Comment by dennisocana** [comment]
In `history` command output the start price and end price seems off in UI.

putty_1LpTxnBand

**Comment by nionis** [comment]
> Hi @nionis I couldn't place orders through Hummingbot even if above the minimum order size.
> 
> putty_ZQfDq0Si6b
> 
> My `order_amount` is `0.001` and the minimum order size for BTC-USD pair is `0.0006`.
> 
> ```
> "pair": "btcusd",
> "price_precision": 5,
> "initial_margin": "20.0",
> "minimum_margin": "10.0",
> "maximum_order_size": "2000.0",
> "minimum_order_size": "0.0006",
> "expiration": "NA",
> "margin": true
> ```
> 
> I can place limit orders in Bitfinex manually with the same amount just fine.
> 
> chrome_LWFAOMyILM
> 
> Log file : [bitfinex_order_amount.log](https://github.com/CoinAlpha/hummingbot/files/4728089/bitfinex_order_amount.log)

Hey, where did you find this data from, I went through the API and I couldn't find "price_precision"

**Comment by dennisocana** [comment]
## Failed to quantize order prices

The order price precision in Hummingbot do not match in exchange / orderbook.

### ETH-USD

image
image
image

### BTC-USD

putty_SewTWcXSD2
chrome_YB3opXcydy

**Comment by dennisocana** [comment]
## Intermittent connection issues

putty_6lZ41ZRpRu

[logs_test_bitfinex.log](https://github.com/CoinAlpha/hummingbot/files/4869254/logs_test_bitfinex.log)

```
2020-07-03 08:07:34,231 - 1798 - hummingbot.market.bitfinex.bitfinex_market - NETWORK - Failed to do order
Traceback (most recent call last):
  File "hummingbot/market/bitfinex/bitfinex_market.pyx", line 532, in hummingbot.market.bitfinex.bitfinex_market.BitfinexMarket._api_do_request
    async with client.request(http_method,
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client.py", line 504, in _request
    await resp.start(conn)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client_reqrep.py", line 847, in start
    message, payload = await self._protocol.read()  # type: ignore  # noqa
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/streams.py", line 591, in read
    await self._waiter
aiohttp.client_exceptions.ClientOSError: [Errno 104] Connection reset by peer
2020-07-03 08:07:34,238 - 1798 - hummingbot.market.bitfinex.bitfinex_market - WARNING - Failed to do order on Bitfinex. Check API key and network connection.
```

## End price in history command

putty_iSjR9BxmSc

**Review by dennisocana — DISMISSED** [review]
Changing review status to approved to merge PR into development. Will create a ticket for the recent bug.

File: hummingbot/market/bitfinex/bitfinex_market.pyx [hunk 1]
```diff
@@ -0,0 +1,1392 @@
...
+                    # amount and price. When amount is 0, it will meant order fill.
+                    if event_type in [ContentEventType.TRADE_UPDATE]:
+                        # amount_come - negative is sell, positive - buy.
+                        # for checking that order is fill
+                        amount_come = Decimal(content["amount"]).quantize(Decimal('1e-8'))
+                        execute_amount_diff = (abs(tracked_order.amount) - abs(amount_come)).quantize(Decimal('1e-8'))
+                        tracked_order.executed_amount_base += abs(amount_come)
+                        tracked_order.executed_amount_quote += abs(amount_come) * execute_price
+
+                    if execute_amount_diff == s_decimal_0.quantize(Decimal('1e-8')) \
```
Comment (vic-en) [inline hummingbot/market/bitfinex/bitfinex_market.pyx:1077]: @nionis 
That line causes the bot to keep executing `_update_order_status` abnormally.
I believe the condition there ought to check if `execute_amount_diff > s_decimal_0`.

**Review by vic-en — DISMISSED** [review]
LGTM

**Comment by dennisocana** [comment]
OK to test

**Comment by dennisocana** [comment]
@nionis please add `bitfinex_api_key` and `bitfinex_secret_key` in `conf_global_TEMPLATE.yml`. See test result below.

```
======================================================================
FAIL: test_global_config_template_complete (test.test_config_templates.ConfigTemplatesUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/jenkins_home/workspace/hb-test_pull-request/test/test_config_templates.py", line 39, in test_global_config_template_complete
    self.assertTrue(key in template_data, f"{key} not in {global_config_template_path}")
AssertionError: False is not true : bitfinex_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml
    'False is not true : bitfinex_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml' = self._formatMessage('False is not true : bitfinex_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true : bitfinex_api_key not in /var/jenkins_home/workspace/hb-test_pull-request/hummingbot/templates/conf_global_TEMPLATE.yml')
```

**Pull request merged**